test/lms_test.c

```diff
@@ -0,0 +1,538 @@
+{
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[1];
+    int ret = 0, i = 0;
+    EVP_PKEY *pkey = NULL;
+    EVP_SIGNATURE *sig = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    unsigned char *pub = NULL;
+    const int step = 1;
+
+    /* Copy the public key data so that we can corrupt it */
+    if (!TEST_ptr(pub = OPENSSL_memdup(td->pub, td->publen)))
+        return 0;
+
+    if (!TEST_ptr(sig = EVP_SIGNATURE_fetch(libctx, "LMS", NULL)))
+        goto end;
+
+    for (i = 0; i < (int)td->publen; i += step) {
+        if (i > 0) {
+            EVP_PKEY_free(pkey);
+            EVP_PKEY_CTX_free(ctx);
+            pkey = NULL;
+            ctx = NULL;
+            pub[i - step] ^= 1;
+        }
+        pub[3] ^= 1;
+        /* Corrupting the public key may cause the key load to fail */
+        pkey = lms_pubkey_from_data(pub, td->publen);
+        if (pkey == NULL)
+            continue;
+        if (!TEST_ptr(ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, NULL)))
+            continue;
+        if (!TEST_int_eq(EVP_PKEY_verify_message_init(ctx, sig, NULL), 1))
+            continue;
+        /* We expect the verify to fail */
+        if (!TEST_int_eq(EVP_PKEY_verify(ctx, td->sig, td->siglen,
+                                         td->msg, td->msglen), 0)) {
+            ret = -1;
+            goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    if (ret == -1)
+        TEST_note("Incorrectly passed when byte %d of the public key"
+                  " was corrupted", i);
+    EVP_SIGNATURE_free(sig);
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_free(pkey);
+    OPENSSL_free(pub);
+
+    return ret == 1;
+}
+
+const OPTIONS *test_get_options(void)
+{
+    static const OPTIONS options[] = {
+        OPT_TEST_OPTIONS_DEFAULT_USAGE,
+        { "config", OPT_CONFIG_FILE, '<',
+          "The configuration file to use for the libctx" },
+        { NULL }
+    };
+    return options;
+}
+
+int setup_tests(void)
+{
+    OPTION_CHOICE o;
+    char *config_file = NULL;
+
+    /* Swap the libctx to test non-default context only */
+    propq = "provider=default";
+
+    while ((o = opt_next()) != OPT_EOF) {
+        switch (o) {
+        case OPT_CONFIG_FILE:
+            config_file = opt_arg();
+            propq = "";
+            break;
+        case OPT_TEST_CASES:
+            break;
+        default:
+        case OPT_ERR:
+            return 0;
+        }
+    }
+    if (!test_get_libctx(&libctx, &nullprov, config_file, &libprov, NULL))
+        return 0;
+
+    ADD_TEST(lms_bad_pub_len_test);
+    ADD_TEST(lms_key_validate_test);
+    ADD_TEST(lms_key_eq_test);
+    ADD_TEST(lms_key_decode_test);
+    ADD_TEST(lms_pubkey_decoder_test);
+    ADD_TEST(lms_pubkey_decoder_fail_test);
+    ADD_ALL_TESTS(lms_verify_test, OSSL_NELEM(lms_testdata));
+    ADD_TEST(lms_verify_fail_test);
+    ADD_TEST(lms_digest_verify_fail_test);
+    ADD_TEST(lms_verify_bad_sig_test);
+    ADD_TEST(lms_verify_bad_sig_len_test);
+    ADD_TEST(lms_verify_bad_pub_sig_test);
```

Comment:
I might have expected to see tests attempting to do key gen, or signature gen.

Reply:
Signature generation wouldn't be a bit difficult considering the nature of this PR, wouldn't it?  I absolutely expect that a signature generation test to show up in the (hopefully upcoming) LMS sign PR

Reply:
This is verify only.  Key gen and signature gen are not implemented.  I don't see the point of a negative test case here.

Reply:
Negative tests added.